pkg/obfuscate/shell.go

```diff
@@ -0,0 +1,607 @@
+
+				index = newIndex
+				obfuscatedSlices = append(obfuscatedSlices, obfuscatedSlicesSubcommand...)
+			}
+		}
+	}
+
+	return obfuscatedSlices
+}
+
+func handleSubcommands(tokens []ShellToken, index int, startIndex int, subcommandType ShellTokenKind) (int, []ObfuscatedSlice) {
+	// We are on a variable or a subcommand
+	// Get the next token to know if that's an open parenthesis and obfuscate the whole subcommand
+	var obfuscatedSlices []ObfuscatedSlice
+
+	openType := ParentheseOpen
+	closingType := ParentheseClose
+
+	if subcommandType == Backticks {
+		openType = Backticks
+		closingType = Backticks
+	}
+
+	// Get all tokens from that subcommand and call this function recursively
+	nbrOpenParentheses := 1
+
+	// Starting index changes depending on the type of subcommand
+	i := startIndex
+
+	for ; i < len(tokens); i++ {
+		if tokens[i].Kind == closingType {
+			nbrOpenParentheses--
+		} else if tokens[i].Kind == openType {
+			nbrOpenParentheses++
+		}
+
+		if nbrOpenParentheses == 0 {
+			// We found the end of the subcommand
+			endSubcommandToken := i
+
+			// Check if at least one token is between the open and close
+			if endSubcommandToken == startIndex {
+				break
+			}
+
+			// Slices of tokens that defines the subcommand
+			subcommandTokens := tokens[startIndex:endSubcommandToken] // without the open and close parentheses
+
+			// Call this function recursively
+			obfuscatedSlices = obfuscateShellCommandToken(subcommandTokens)
+
+			// Increment the index to the end of the subcommand
+			index = endSubcommandToken
+			break
+		}
+	}
+
+	return index, obfuscatedSlices
+}
+
+// grabFullArgument grabs the full argument from the given tokens starting at the given index.
+// returns the index of the last token of the argument and the number of tokens grabbed.
+// For examples:
+// - if the argument starting at index 3 is a Dollar token, then it should also grab the next token (calling itself recursively) as part of the argument.
+// - if the argument starting at index 3 is a Field token, then it should only grab the Field token.
+// - if the argument starting at index 3 is an Equal token, then it should grab the Equal token and the next token (calling itself recursively).
+// - if the argument starting at index 3 is a SingleQuote or DoubleQuote token, then it should grab the whole quoted string.
+func grabFullArgument(tokens []ShellToken, index int) (int, int) {
+	tokensLength := len(tokens)
+	if index >= tokensLength {
+		return tokensLength - 1, 0 // Return the last token index if we are out of bounds
+	}
+
+	nbrGrabbed := 0
+	for ; index < tokensLength; index++ {
+		kind := tokens[index].Kind
+
+		// We can't grab a control or redirection token
+		if kind == Control || kind == Redirection {
+			return index - 1, nbrGrabbed
+		}
+
+		// Grab only the current token
+		if kind == Field || kind == Equal || kind == ShellVariable {
+			return index, nbrGrabbed + 1
+		}
+
+		// Grab the next token
+		if kind == Dollar || kind == Equal {
```

Comment:
kind == Equal can never be true here as it's also compared against on 422 (it seems like it shouldn't be on 422 there since an Equal seems like it's supposed to not just grab the current token, Is there a missing test we can add that will catch this, or is the behavior similar enough that the tests don't fail here?)

Reply:
Related: maybe refactor this to be a switch to make it easier to find these sorts of things? (not strictly required but it might improve readability here)